### PR TITLE
fix: update certificate handling RFC.

### LINKED ISSUE #39: Update certificate handling RFC
Update the [certificate handling RFC](https://github.com/kubewarden/rfc/blob/main/rfc/0018-certificate-handling.md) with the actual implementation.

The description of these issues could be used as a reference:

kubewarden/kubewarden-controller#820 
kubewarden/kubewarden-controller#819
 kubewarden/kubewarden-controller#818

### PR DESCRIPTION
**Pull request by jvanz**
## Description

Updates the RFC 18 about certificate handling to reflect what have been done during the removal of the CertManager.

Fix https://github.com/kubewarden/rfc/issues/39

### CODE REVIEW

File: rfc/0018-certificate-handling.md [hunk 1]
```diff
@@ -82,174 +86,195 @@ webhooks:
...
-- Create Policy Server Deployment, ensure the certificate is mounted inside of the Pods
-
-This is already done by our current code (Kubewarden 1.6.0).
+- Create Policy Server Deployment, ensure the certificate is mounted inside of
+  the Pods
 
-### New `AdmissionPolicy`/`ClusterAdmissionPolicy` is defined
+It's important to note that the CA used to sign the Policy Server certificate
+is the same CA used to sign the `kubewarden-controller` certificate. It's not
+the internal CA generated to signed Policy Server certificates as it is done
```
Comment (flavio) [inline rfc/0018-certificate-handling.md]:
```suggestion
the internal CA generated to signed Policy Server certificates as it was done
```

File: rfc/0018-certificate-handling.md [hunk 2]
```diff
@@ -82,174 +86,195 @@ webhooks:
...
 
 - Generate new certificate, sign it with our CA
-- Force a rollout of the Policy Server deployment. This will ensure the HTTPS server uses the new certificate
-
-> **Note:** there's no need to touch the webhook configuration objects related to the policies
-> hosted by the Policy Server. The new certificate is still signed by the same CA, hence nothing has
+- Update the secret storing the certificate
+- Once Kubernetes updates the certificate secret mount inside of the Pod, the
+  Policy Server or `kubewarden-controller` will detect the change in the
+  certificate file and start using the new certificate
```
Comment (flavio) [inline rfc/0018-certificate-handling.md]:
```suggestion
- Kubernetes automatically updates the certificate secret mounted inside of the Pods
  that are running the Policy Server or the `kubewarden-controller`. Both types of
  running process will notice the change of the certificate file they read at
  boot time and will reload their TLS configuration.
```

File: rfc/0018-certificate-handling.md [hunk 3]
```diff
@@ -82,174 +86,195 @@ webhooks:
...
-  - Update the `clientConfig.caBundle`: ensure it contains only the certitificate of the new CA
+- For each webhook configuration managed by Kubewarden (meaning all the
+  policies deployed, plus the `kubewarden-controller`):
+  - Update the `clientConfig.caBundle`: ensure it contains only the
+    certitificate of the new CA
 
-By respecing these steps we can avoid communication failures between the Kubernetes API server and the webhooks.
+> **Note:** there's no need to restart the Policy Server Pod. Both Policy
+> Server and the `kubewarden-controller` are able to detect changes in the
+> certificate file and reload it on runtime.
```
Comment (flavio) [inline rfc/0018-certificate-handling.md]:
```suggestion
The root CA rotation is initiated 60 days before the expiration date.
This gives plenty of time to ensure all the Policy Server instances and the
`kubewarden-controller` pods are using a certificate that has been issued
by the new root CA.

Once the old CA is invalid:
- Remove the old root CA from the CA bundle
- For each webhook configuration managed by Kubewarden (meaning all the
  policies deployed, plus the validating and mutating webhooks
  defined by the Kubewarden Controller):
  - Update the `clientConfig.caBundle`: use the freshly computed bundle
  
> **Note:** there's no need to restart the Policy Server/kubewarden-controller pods.
> By that time (60 days passed since the new root CA and their certificates were
> recreated), they are all using a certificate issued by the new root CA.
```

File: rfc/0018-certificate-handling.md [hunk 4]
```diff
@@ -17,8 +18,8 @@ having to resort to external dependencies like CertManager.
 # Motivation
 
 The first goal of this RFC is to provide a picture of the current state of
-certificate handling inside of Kubewarden. How many certificates are used,
-by which components, which CA signed them and how have they been generated?
+certificate handling inside of Kubewarden. How many certificates are used, by
```
Comment (jhkrug) [inline rfc/0018-certificate-handling.md]:
```suggestion
certificate handling in Kubewarden. How many certificates are used, by
```

File: rfc/0018-certificate-handling.md [hunk 5]
```diff
@@ -35,38 +36,41 @@ certificate rotation.
 
 # Detailed design
 
+Let's start by describing the current state of certificate handling inside of
```
Comment (jhkrug) [inline rfc/0018-certificate-handling.md]:
```suggestion
Let's start by describing the current state of certificate handling in
```

File: rfc/0018-certificate-handling.md [hunk 6]
```diff
@@ -35,38 +36,41 @@ certificate rotation.
...
 
 ## Webhook Configuration and Certificates
 
 Both `ValidatingWebhookConfiguration` and `MutatingWebhookConfiguration`
 resources have a
 [`clientConfig`](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.27/#webhookclientconfig-v1-admissionregistration-k8s-io)
-object inside of their definition.
-The `clientConfig.caBundle` has the bundle of CA required to
-verify the TLS termination of the webhook endpoint:
+object inside of their definition. The `clientConfig.caBundle` has the bundle
```
Comment (jhkrug) [inline rfc/0018-certificate-handling.md]:
```suggestion
object in their definition. The `clientConfig.caBundle` has the bundle
```

File: rfc/0018-certificate-handling.md [hunk 7]
```diff
@@ -35,38 +36,41 @@ certificate rotation.
...
+of CA required to verify the TLS termination of the webhook endpoint:
```
Comment (jhkrug) [inline rfc/0018-certificate-handling.md]:
```suggestion
of CAs required to verify the TLS termination of the webhook endpoint:
```

File: rfc/0018-certificate-handling.md [hunk 8]
```diff
@@ -82,174 +86,199 @@ webhooks:
...
-- Start the Kubewarden controller HTTPS server using this certificate
-- Register the webhook against the Kubernetes API, using the certificate of our CA inside of
-  `clientConfig.caBundle`
+- Helm chart generates Certificate Authority(CA) and store it in a secret
+- Helm chart generates cerficate for `kubewarden-controller`, sign it with the
+  CA, and store it in a secret
+- Helm chart uses the CA in the `caBundle` configuration in webhooks to
+  validate CRDs pointing to the `kubewarden-controller`
+- Start the Kubewarden controller HTTPS server using this certificate. The
+  certificate is mounted inside of the Pod
```
Comment (jhkrug) [inline rfc/0018-certificate-handling.md]:
```suggestion
- Helm chart generates Certificate Authority(CA) and stores it in a secret
- Helm chart generates certificate for `kubewarden-controller`, signs it with the
  CA, and stores it in a secret
- Helm chart uses the CA in the `caBundle` configuration in webhooks to
  validate CRDs pointing to the `kubewarden-controller`
- Start the Kubewarden controller HTTPS server using this certificate. The
  certificate is mounted inside the Pod
```

File: rfc/0018-certificate-handling.md [hunk 9]
```diff
@@ -82,174 +86,199 @@ webhooks:
...
 
 ### New Policy Server Defined
 
 A new `PolicyServer` is defined, the following actions have to be done:
 
 - Generate new certificate, sign it with our CA
-- Create Policy Server Deployment, ensure the certificate is mounted inside of the Pods
+- Create Policy Server Deployment, ensure the certificate is mounted inside of
+  the Pods
```
Comment (jhkrug) [inline rfc/0018-certificate-handling.md]:
```suggestion
- Create Policy Server Deployment, ensure the certificate is mounted in
  the Pods
```

File: rfc/0018-certificate-handling.md [hunk 10]
```diff
@@ -82,174 +86,199 @@ webhooks:
...
 
-This is already done by our current code (Kubewarden 1.6.0).
+It's important to note that the CA used to sign the Policy Server certificate
+is the same CA used to sign the `kubewarden-controller` certificate. It's not
+the internal CA generated to signed Policy Server certificates as it was done
+before.
```
Comment (jhkrug) [inline rfc/0018-certificate-handling.md]:
```suggestion
It's important to note that the CA used to sign the Policy Server certificate
is the same CA used to sign the `kubewarden-controller` certificate. It's not
the internal CA generated to sign Policy Server certificates as was done
before.
```

File: rfc/0018-certificate-handling.md [hunk 11]
```diff
@@ -82,174 +86,199 @@ webhooks:
...
 
-### New `AdmissionPolicy`/`ClusterAdmissionPolicy` is defined
+### New
+`AdmissionPolicy`/`ClusterAdmissionPolicy`/`ClusterAdmissionPolicyGroup`/`AdmissionPolicyGroup`
+is defined
 
-A new policy is defined inside of the cluster, ensure the following action are done:
+A new policy is defined inside of the cluster, ensure the following action are
+done:
```
Comment (jhkrug) [inline rfc/0018-certificate-handling.md]:
```suggestion
A new policy is defined inside the cluster, ensure the following actions are
done:
```

File: rfc/0018-certificate-handling.md [hunk 12]
```diff
@@ -82,174 +86,199 @@ webhooks:
...
 
-- Register the webhook against the Kubernetes API, use our CA inside of `clientConfig.caBundle`
-
-> **Note:** it doesn't matter which Policy Server is going to host the policy. All the Policy Server
-> instances use a certificate that is signed by our CA. Hence we always put our CA
-> inside of the `clientConfig.caBundle`.
+- Register the webhook against the Kubernetes API, use our CA inside of
+  `clientConfig.caBundle`
```
Comment (jhkrug) [inline rfc/0018-certificate-handling.md]:
```suggestion
- Register the webhook against the Kubernetes API, use our CA in
  `clientConfig.caBundle`
```

File: rfc/0018-certificate-handling.md [hunk 13]
```diff
@@ -82,174 +86,199 @@ webhooks:
...
 
-This is already done by our current code (Kubewarden 1.6.0).
+> **Note:** it doesn't matter which Policy Server is going to host the policy.
+> All the Policy Server instances use a certificate that is signed by our CA.
+> Hence we always put our CA inside of the `clientConfig.caBundle`.
```
Comment (jhkrug) [inline rfc/0018-certificate-handling.md]:
```suggestion
> Hence we always put our CA in the `clientConfig.caBundle`.
```

File: rfc/0018-certificate-handling.md [hunk 14]
```diff
@@ -82,174 +86,199 @@ webhooks:
...
 
-### Policy Server Certificate Renewed
+### Policy Server and `kubewarden-controller` Certificates Renewed
 
-From time to time we need to renew the certificate issued to a Policy Server instance. The main reason
-to perform this operation is to ensure the Policy Server is not using an expired certificate.
+From time to time we need to renew the certificate issued to the
+`kubewarden-controller` and a Policy Server instance. The main reason to
+perform this operation is to ensure the certicates in use are not expired.
```
Comment (jhkrug) [inline rfc/0018-certificate-handling.md]:
```suggestion
perform this operation is to ensure the certificates in use do not expire.
```

File: rfc/0018-certificate-handling.md [hunk 15]
```diff
@@ -82,174 +86,199 @@ webhooks:
...
 
-To renew a Policy Server certificate the following actions have to be done:
+To renew a certificate the following actions have to be done:
```
Comment (jhkrug) [inline rfc/0018-certificate-handling.md]:
```suggestion
To renew a certificate the following actions are needed:
```

File: rfc/0018-certificate-handling.md [hunk 16]
```diff
@@ -82,174 +86,199 @@ webhooks:
...
 - Generate new certificate, sign it with our CA
-- Force a rollout of the Policy Server deployment. This will ensure the HTTPS server uses the new certificate
-
-> **Note:** there's no need to touch the webhook configuration objects related to the policies
-> hosted by the Policy Server. The new certificate is still signed by the same CA, hence nothing has
+- Update the secret storing the certificate
+- Kubernetes automatically updates the certificate secret mounted inside of the
+  Pods that are running the Policy Server or the `kubewarden-controller`. Both
+  types of running process will notice the change of the certificate file they
+  read at boot time and will reload their TLS configuration.
```
Comment (jhkrug) [inline rfc/0018-certificate-handling.md]:
```suggestion
- Kubernetes automatically updates the certificate secret mounted inside the
  Pods that are running the Policy Server or the `kubewarden-controller`. Both
  types of running process notice the change of the certificate file they
  read at boot time and will reload their TLS configuration.
```

File: rfc/0018-certificate-handling.md [hunk 17]
```diff
@@ -82,174 +86,199 @@ webhooks:
...
+
+> **Note:** there's no need to restart the Pods. Since Kubewarden v1.16, Policy
+> Server and `kubewarden-controller` are able to detect changes in the
+> certificate file and reload it on runtime. Also, there's no need to touch the
```
Comment (jhkrug) [inline rfc/0018-certificate-handling.md]:
```suggestion
> certificate file and reload it at runtime. Also, there's no need to touch the
```

File: rfc/0018-certificate-handling.md [hunk 18]
```diff
@@ -82,174 +86,199 @@ webhooks:
...
-controller creates the root CA only when the 1st PolicyServer object is defined.
-We can keep creating a root CA that expires after 10 years.
-
-The code must also ensure the root CA is not about to expire. When this scenario is detected, the controller must renew
-the root CA.
+Helm chart installation is responsible to create the first root CA. The root CA
+is used to sign the first certificate user by the `kubewarden-controller`. The
+root CA is also used in the `clientConfig.caBundle` of the webhooks used to
+validate the CRDs. The root CA has a validity of 10 years and the certificate
+used by the controller has a validity of 1 year. Both uses the ECDSA P-256
```
Comment (jhkrug) [inline rfc/0018-certificate-handling.md]:
```suggestion
used by the controller has a validity of 1 year. Both use the ECDSA P-256
```

File: rfc/0018-certificate-handling.md [hunk 19]
```diff
@@ -82,174 +86,199 @@ webhooks:
...
-
-- `PolicyServer`
-- `ClusterAdmissionPolicy`
-- `AdmissionPolicy`
-
-### Policy Server
+This section describes how the `kubewarden-controller` reconciliation loops
+should be arranged. We can split the changes in three different parts: Policy
+Server reconciliation loop changes; Policy types reconciliation loop changes;
+New certificates reconciliation loop;
```
Comment (jhkrug) [inline rfc/0018-certificate-handling.md]:
```suggestion
This section describes how the `kubewarden-controller` reconciliation loops
should be arranged. We can split the changes in three different parts: 

- Policy Server reconciliation loop changes
- Policy types reconciliation loop changes
- New certificates reconciliation loop
```

File: rfc/0018-certificate-handling.md [hunk 20]
```diff
@@ -82,174 +86,199 @@ webhooks:
...
 
-> Note: checking if the Policy Server certificate has been issued by our CA might be computationally intense.
-> We have to ensure this is not going to cause a CPU spike. We could resort to do something clever using
-> the revision number of the Secret that holds the CA.
+If during the reconciliation loop the root CA is about to expire expired. In
+other words, if the certificate will expire in less than 60 days, the
+reconciliation loop should generate a new root CA, update the secret storing
+the root CA and update the `clientConfig.caBundle` of all the webhook
+configurations. And then it should generate a new certificate for the
+`kubewarden-controller` and all the policy server deployed in the cluster.
```
Comment (jhkrug) [inline rfc/0018-certificate-handling.md]:
```suggestion
If during the reconciliation loop the root CA is about to expire, that is, 
the certificate will expire in less than 60 days, the
reconciliation loop should generate a new root CA, update the secret storing
the root CA and update the `clientConfig.caBundle` of all the webhook
configurations. Then it should generate a new certificate for the
`kubewarden-controller` and all the policy servers deployed in the cluster.
```

File: rfc/0018-certificate-handling.md [hunk 21]
```diff
@@ -82,174 +86,199 @@ webhooks:
...
 
-### ClusterAdmissionPolicy
+Also during the reconciliation, if a certificate used by a Policy Server or the
+`kubewarden-controller` is about to expire, the reconciliation loop should
+generate a new certificate, singed it with the root CA and update the secret
```
Comment (jhkrug) [inline rfc/0018-certificate-handling.md]:
```suggestion
generate a new certificate, sign it with the root CA and update the secret
```

File: rfc/0018-certificate-handling.md [hunk 22]
```diff
@@ -82,174 +86,199 @@ webhooks:
...
+storing the certificate. There is no need to update the webhook configurations,
+since the CA is the same.
 
-The reconciliation loop should watch also the Secret object that contains the root CA certificate.
+If the old CA stored together in the secret with the current CA in use is
+invalid, the reconciliation loop should update the secret removing the old CA
+and then update the `clientConfig.caBundle` of all the webhook configurations
+to contain only the new CA.
```
Comment (jhkrug) [inline rfc/0018-certificate-handling.md]:
```suggestion
If the old CA stored in the secret with the current CA in use is
invalid, the reconciliation loop should update the secret, removing the old CA.
Then it should update the `clientConfig.caBundle` of all webhook configurations
to contain only the new CA.
```

**Review by jhkrug — CHANGES_REQUESTED** [review]
Very minor suggestions, and a couple or three typos.

File: rfc/0018-certificate-handling.md [hunk 23]
```diff
@@ -82,174 +86,199 @@ webhooks:
...
-flowchart TD
-    A[KW Internal CA]
-    A -->B(KW controller)
-    A -->C(Policy Server #1)
-    A -->D(Policy Server #2)
-```
+We would like to get rid of CertManager. To do that we need to change how
+certificates are managed for the `kubewarden-controller`. Going forward, the
+Helm chart installation and  the controller will take care of generating the
+certificate used by the `kubewarden-controller`.
```
Comment (fabriziosestito) [inline rfc/0018-certificate-handling.md]:
```suggestion
certificates are managed for the `kubewarden-controller`. Going forward, the
Helm chart installation and  the controller will take care of bootstrapping the
certificates, while the `kubewarden-controller` will rotate them when they are about to expire.
```

File: rfc/0018-certificate-handling.md [hunk 24]
```diff
@@ -82,174 +86,199 @@ webhooks:
...
+The Kubewarden stack has just been installed. The following actions have to be
+done:
 
-- Generate Certificate Authority
-- Generate certificate for `kubewarden-controller` and sign it with our CA
-- Start the Kubewarden controller HTTPS server using this certificate
-- Register the webhook against the Kubernetes API, using the certificate of our CA inside of
-  `clientConfig.caBundle`
+- Helm chart generates Certificate Authority(CA) and store it in a secret
+- Helm chart generates cerficate for `kubewarden-controller`, sign it with the
```
Comment (fabriziosestito) [inline rfc/0018-certificate-handling.md]:
```suggestion
Helm chart generates a certificate for the `kubewarden-controller` webhooks, sign it with the
```

File: rfc/0018-certificate-handling.md [hunk 25]
```diff
@@ -82,174 +86,199 @@ webhooks:
...
+  CA, and store it in a secret
+- Helm chart uses the CA in the `caBundle` configuration in webhooks to
+  validate CRDs pointing to the `kubewarden-controller`
+- Start the Kubewarden controller HTTPS server using this certificate. The
```
Comment (fabriziosestito) [inline rfc/0018-certificate-handling.md]:
```suggestion
Start the Kubewarden controller webhooks HTTPS server using this certificate. The
```

File: rfc/0018-certificate-handling.md [hunk 26]
```diff
@@ -82,174 +86,199 @@ webhooks:
...
+  certificate is mounted inside of the Pod
 
 ### New Policy Server Defined
 
 A new `PolicyServer` is defined, the following actions have to be done:
 
 - Generate new certificate, sign it with our CA
-- Create Policy Server Deployment, ensure the certificate is mounted inside of the Pods
+- Create Policy Server Deployment, ensure the certificate is mounted inside of
```
Comment (fabriziosestito) [inline rfc/0018-certificate-handling.md]: We don't ensure that a certificate is mounted anymore. We leverage an auto-reload mechanism since the secret volume will eventually mounted/updated by Kubernetes.

File: rfc/0018-certificate-handling.md (hunk 14 above)
Comment (fabriziosestito) [inline rfc/0018-certificate-handling.md]: ~when do we do this? we can mention the lookahead time~

nevermind I see it was added to the section below

File: rfc/0018-certificate-handling.md (hunk 16 above)
Comment (fabriziosestito) [inline rfc/0018-certificate-handling.md]: not at boot time as we are auto-realoading in both components

**Review by fabriziosestito — CHANGES_REQUESTED** [review]

File: rfc/0018-certificate-handling.md (hunk 16 above)
Comment (jvanz) [reply rfc/0018-certificate-handling.md]: Yes, that's what the sentence is trying to say in a different way. Let me rephrase it to make it more clear.

File: rfc/0018-certificate-handling.md [hunk 27]
```diff
@@ -82,174 +86,201 @@ webhooks:
...
-created. Check the previous section to see which actions have to be done in this circumstance.
+### Certificates Reconciliation Loop
 
-The loop should also monitor changes done to the secret that holds the internal CA.
-The reconciliation loop should then make sure that the Policy Server certificate has been signed by
-our CA. If not, a new certificate should be issued (like during a renewal).
+The new added reconciliation loop should check the expiration date of all the
+certificates used by the Kubewarden stack. Actually, this reconciliation loop
+is a runnable added to the controller manager. The runnable starts a ticker
+that verify validity of the certificates every 12 hours.
```
Comment (jhkrug) [inline rfc/0018-certificate-handling.md]:
```suggestion
that verifies validity of the certificates every 12 hours. 
```

**Review by jhkrug — CHANGES_REQUESTED** [review]
One tiny typo otherwise lgtm.

**Review by fabriziosestito — APPROVED** [review]

**Review by viccuad — APPROVED** [review]
LGTM modulo John's suggestion.
